blog/tidy-collaborative-filtering/index.qmd

```diff
@@ -0,0 +1,216 @@
+  "Item-Based Filtering" = list(name  = "IBCF", param = list(k = 10)),
+  "User-Based Filtering" = list(name  = "UBCF", param = list(method = "Cosine", nn = 100))
+)
+
+number_of_recommendations <- c(1, 3, 5, 10)
+results <- evaluate(
+  scheme,
+  algorithms,
+  type = "topNList",
+  progress = TRUE,
+  n = number_of_recommendations
+)
+```
+
+```{r eval = FALSE, include = FALSE}
+write_rds(results, "../../data/results.rds")
+```
+
+```{r include = FALSE}
+results <- read_rds("../../data/results.rds")
+```
+
+# Evaluate Recommenders
+
+The output of `evaluate()` already provides the evaluation metrics in a structured way. I can simply average the metrics over the cross-validation folds.
+
+```{r}
+results_tbl <- results |>
+  avg() |>
+  map(as_tibble) |>
+  bind_rows(.id = "model")
+```
+
+Now, for each recommender, we are interested the following numbers:
+
+* True Negative (TN) = number of not predicted items that do not correspond to withheld items
+* False Positive (FP) = number of incorrect predictions that do not correspond to withheld items
+* False negative (FN) = number of not predicted items that correspond to withheld items
+* True Positive (TP) = number of correct predictions that correspond to withheld items
+
+The two figures below present the most common evaluation techniques for the performance of recommender algorithms in backtest settings like mine. 
+
+## ROC curves
+
+The first visualization approach comes from signal-detection and is called "Receiver Operating Characteristic" (ROC). The ROC-curve plots the algorithm’s probability of detection (TPR) against the probability of false alarm (FPR).
+
+* TPR = TP / (TP + FN) (i.e. share of true positive recommendations relative to all known portfolios)
+* FPR = FP / (FP + TN) (i.e. share of incorrect recommendations relative to )
+
+Intuitively, the bigger the area under the ROC curve, the better is the corresponding algorithm.
+
+```{r}
+#| fig-alt: "Title: ROC curves of stock recommender algorithms. The figure shows that the association rules mining algorithm achieves the highest true positive rate for any false positive rate."
+results_tbl |>
+  ggplot(aes(FPR, TPR, colour = model)) +
+  geom_line() +
+  geom_label(aes(label = n))  +
+  labs(
+    x = "False Positive Rate (FPR)",
+    y = "True Positive Rate (TPR)",
+    title = "ROC curves of stock recommender algorithms",
+    colour = "Model"
+  ) +
+  theme(legend.position = "right") + 
+  scale_y_continuous(labels = percent) +
+  scale_x_continuous(labels = percent)
+```
+
+The figure shows that recommending random items exhibits the lowest TPR for any FPR, so it is the worst among all algorithms (which it is not surprising). Association rules, on the other hand, constitutes the best algorithm among the current selection. This result is neat because association rule mining is a computationally cheap algorithm, so we could potentially fine-tune or reestimate the model easily.
```

Comment:
**Typos**:
"(which is not surprising)"
"Association rules, on the other hand, constitute"